design/tls-client-verification.md

```diff
@@ -0,0 +1,143 @@
+# Client certificate validation (mTLS)
+
+Status: Draft
+
+## Goals
+
+- Allow client certificate validation (mTLS) on contour (performed by Envoy)
+- Allow various ways for client certificate validation; Spki, Hash and CA
```

Comment:
Can only one type be specified? Envoy allows multiple, but I'm not sure of the use-cases that would drive picking multiple.

> When both: verify_certificate_hash and verify_certificate_spki are specified, a hash matching value from either of the lists will result in the certificate being accepted.

Reply:
Me neither to be honest. I rely on that Envoy knows what they are doing and try to allow all Envoy configurations for mTLS also in `contour`. That is also a reason for the non-goal;
> Repeat configuration details described in Envoy documentation (make references)
I will pass the question on to a security expert, but they tend to be busy...

Reply:
I think using multiple types makes most sense in combination of a `trusted_ca` and `verify_subject_alt_name`. In this case the administrator wants to uniquely authenticate a particular client(s) before allowing them to establish TLS connection.  For example a client with identity "CN=Joe" issued by certain trusted CA (or a sub-CAs of the trusted CA).  This is by far the most common use case for client certificates in general.  However, like discussed in PR comments, in some cases the specific client identity might not matter and one would then specify  `trusted_ca` alone.

Combination of `verify_certificate_spki` and `verify_certificate_hash` does not make sense to me, since it is likely that administrator picks just one way to pin a specific certificate: either administrator wants to identify client(s) by hash of public key or hash of the complete certificate file, not both at the same time.  Since this is an act of whitelisting a specific client certificate, it would be bit redundant or at least confusing, to use `trusted_ca` in combination with either of these two methods. When a client identified by a specific cert is explicitly marked as allowed to establish a TLS connection, validating that it is also issued by a trusted CA should not generally matter (though I have not tested if this combination is supported by Envoy). Obviously Envoy will still require the client to demonstrate proof-of-possession of the private key associated to the client certificate, in addition to validating the hash of the public part.

In theory one could imagine a use case where some clients are allowed to establish a connection because they are issued by `trusted_ca` and others because they are specifically pinned/whitelisted by their `verify_certificate_spki` and `verify_certificate_hash` fingerprint but I don't think this is very common scenario.

Reply:
@tsaarni I have ignored `verify_subject_alt_name` since I did not understand it's use. From your comment I see it's important. Can the configuration look something like this?;
```
      clientValidation:
        secretName: clientsecret
        subjectAltNames:
          - Joe
```
The list of `subjectAltNames` would be passed to Envoy in the `verify_subject_alt_name` field.

Reply:
Envoy doc; https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/auth/cert.proto.html?highlight=require_client_certificate#auth-certificatevalidationcontext

Reply:
Correction to my comment:  My example "CN=Joe" (X509 Common Name field) was for subjectName and subjectAltName would rather be something like email "joe@example.com", instead of just "joe".   Obvious from documentation, Envoy only supports validation of subjectAltName field. 

One complication is that X509 subjectAltName is not just free text: there is type involved (see list [here]( https://en.wikipedia.org/wiki/Subject_Alternative_Name) and definition of [GeneralName in RFC5280]( https://tools.ietf.org/html/rfc5280#section-4.2.1.6).  However Envoy documentation does not seem to specify what format it supports?

I have small doubt that maybe it just supports URIs since this feature might be added to Envoy due to Istio, which uses subjectAltName in the [SPIFFE certs](https://github.com/spiffe/spiffe/blob/master/standards/X509-SVID.md) and it leaves subjectName empty.   I might try to find this from envoy code.

Reply:
Ok thanks for the update. I will change from "Joe" to an URI, "service.example.com".

Reply:
Just out of curiosity I checked what types of subjectAltName's Envoy supports as values of `verify_subject_alt_name`.  The code is [here](https://github.com/envoyproxy/envoy/blob/81654d18dfe211d739082c922776cf18bbf03275/source/extensions/transport_sockets/tls/context_impl.cc#L489).   The value can be DNS name, URI or IPv4/6 address.